org/mozilla/jss/ssl/javax/JSSEngine.java

```diff
@@ -905,6 +917,58 @@ public boolean getWantClientAuth() {
         return want_client_auth;
     }
 
+    /**
+     * Set a specific list of protocols to negotiate next for ALPN support.
+     */
+    public void setApplicationProtocols(String[] protocols) {
+        alpn_protocols = protocols;
+    }
+
+    /**
+     * Get the most recently negotiated application protocol.
+     *
+     * Note that NSS only allows selection on the initial handshake so
+     * this is implemented via a call to getHandshakeApplicationProtocol().
+     */
+    public String getApplicationProtocol() {
+        return getHandshakeApplicationProtocol();
+    }
+
+    /**
+     * Get the application protocol negotiated during the initial handshake.
+     */
+    public String getHandshakeApplicationProtocol() {
+        if (session == null) {
+            return null;
+        }
+
+        return session.getNextProtocol();
+    }
+
+    /**
+     * Helper method for implementations: encodes ALPN protocols into wire
+     * format (8-bit length prefixed byte encoding).
+     */
+    public byte[] getALPNWireData() {
+        int length = 0;
+        for (String protocol : alpn_protocols) {
+            length += 1 + protocol.getBytes(StandardCharsets.UTF_8).length;
+        }
+
+        byte[] result = new byte[length];
+        int offset = 0;
+
+        for (String protocol : alpn_protocols) {
+            byte[] p_bytes = protocol.getBytes(StandardCharsets.UTF_8);
+            result[offset] = (byte) p_bytes.length;
+            offset += 1;
+            System.arraycopy(p_bytes, 0, result, offset, p_bytes.length);
+            offset += p_bytes.length;
+        }
+
+        return result;
```

Comment:
We should also check that `result.length < 65536`.  Not sure what exception we should raise if that were not met, but we should check it.